src/widget/wbasicpushbutton.cpp

```diff
@@ -0,0 +1,115 @@
+#include "widget/wbasicpushbutton.h"
+
+#include <QPaintEvent>
+#include <QStyleOption>
+#include <QStylePainter>
+#include <QToolTip>
+
+#include "moc_wbasicpushbutton.cpp"
+
+namespace {
+void elideButtonLabelIfNeeded(QStyleOptionButton& option,
+        QStyle* style,
+        QWidget* widget,
+        Qt::TextElideMode elideMode) {
+    if (elideMode == Qt::ElideNone) {
+        return;
+    }
+
+    // Calculate how much space we have left for the text
+    QSize textSize = style->subElementRect(QStyle::SE_PushButtonContents, &option, widget).size();
+
+    if (option.features & QStyleOptionButton::HasMenu) {
+        int indicatorSize = style->pixelMetric(QStyle::PM_MenuButtonIndicator, &option, widget);
+        textSize.setWidth(textSize.width() - indicatorSize);
+    }
+
+    if (!option.icon.isNull()) {
+        int iconSpacing = 4; // ### 4 is currently hardcoded in QPushButton::sizeHint()
+        textSize.setWidth(textSize.width() - option.iconSize.width() - iconSpacing);
+    }
+
+    // Replace overflowing text with ellipsis ("...")
+    option.text = option.fontMetrics.elidedText(option.text, elideMode, textSize.width());
+}
+} // namespace
+
+WBasicPushButton::WBasicPushButton(QWidget* pParent)
+        : QPushButton(pParent),
+          m_elideMode(Qt::ElideNone) {
+}
+
+void WBasicPushButton::setElideMode(Qt::TextElideMode elideMode) {
+    if (elideMode != m_elideMode) {
+        m_elideMode = elideMode;
+        updateGeometry();
+    }
+}
+
+QSize WBasicPushButton::minimumSizeHint() const {
+    QSize fullSize = sizeHint();
+
+    if (m_elideMode == Qt::ElideNone) {
+        // Elision of overflowing text is disabled,
+        // so this button cannot be collapsed beyond
+        // its default size.
+        return fullSize;
+    } else {
+        // Elision of overflowing text is enabled.
+        return QSize(0, fullSize.height());
+    }
+}
+
+void WBasicPushButton::paintEvent(QPaintEvent* /*unused*/) {
+    QStylePainter p(this);
+    QStyleOptionButton option;
+    initStyleOption(&option);
+    elideButtonLabelIfNeeded(option, p.style(), this, m_elideMode);
+    p.drawControl(QStyle::CE_PushButton, option);
+}
+
+QString WBasicPushButton::buildToolTip() const {
+    // Show the button label as a tooltip when the label text
+    // is partially hidden due to the button's size
+
+    QString txtToolTip = toolTip();
+
+    if (!txtToolTip.isEmpty()) {
+        return txtToolTip;
+    }
+
+    QString txtLabel = text();
+    QSize currentSize = size();
+    QSize fullSize = sizeHint();
+
+    if ((currentSize.height() < fullSize.height() ||
+                currentSize.width() < fullSize.width()) &&
+            !txtLabel.isEmpty()) {
```

Comment:
Why do we need to check the height here?

Reply:
It is possible that there is enough horizontal space but not enough vertical space to display the full button text. The simplest case I can think of is when `QPushButton::text()` contains linebreaks, but the button height is restricted to a single line.

The fallback tooltip should be enabled in all cases where the button text is not fully visible, which applies here. (Sidenote: It is not enabled when the button text is fully visible to avoid "useless"/confusing tooltips that only duplicate the information shown on the button itself.)

Reply:
I think we should implement only the cases we need currently and expand as soon as we need multi-line buttons.
IMHO the context should be clear so that we don't need multi-line buttons in the first place.
Or do you have an exmple where we might need one?